src/Components/Components/src/Infrastructure/ComponentStatePersistenceManager.cs

```diff
@@ -44,47 +52,48 @@ public async Task RestoreStateAsync(IPersistentComponentStateStore store)
     /// <summary>
     /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>.
     /// </summary>
-    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to restore the application state from.</param>
+    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to persist the application state into.</param>
     /// <param name="renderer">The <see cref="Renderer"/> that components are being rendered.</param>
     /// <returns>A <see cref="Task"/> that will complete when the state has been restored.</returns>
     public Task PersistStateAsync(IPersistentComponentStateStore store, Renderer renderer)
         => PersistStateAsync(store, renderer.Dispatcher);
 
     /// <summary>
-    /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>.
+    /// Persists the component application state into the given <see cref="IPersistentComponentStateStore"/>
+    /// so that it could be restored on Server.
     /// </summary>
-    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to restore the application state from.</param>
+    /// <param name="store">The <see cref="IPersistentComponentStateStore"/> to persist the application state into.</param>
     /// <param name="dispatcher">The <see cref="Dispatcher"/> corresponding to the components' renderer.</param>
     /// <returns>A <see cref="Task"/> that will complete when the state has been restored.</returns>
     public Task PersistStateAsync(IPersistentComponentStateStore store, Dispatcher dispatcher)
     {
-        if (_stateIsPersisted)
-        {
-            throw new InvalidOperationException("State already persisted.");
-        }
-
-        _stateIsPersisted = true;
-
         return dispatcher.InvokeAsync(PauseAndPersistState);
 
         async Task PauseAndPersistState()
         {
-            State.PersistingState = true;
-            await PauseAsync();
-            State.PersistingState = false;
+            var currentState = new Dictionary<string, byte[]>();
 
-            await store.PersistStateAsync(_currentState);
+            State.PersistenceContext = new(currentState);
+            await PauseAsync(store);
+            State.PersistenceContext = default;
```

Comment:
We can avoid all this if we do things a bit differently.

* Collect or infer the mode inside the call to register for each callback.
* Persist components with Server mode in a batch.
* Persist components with Webassembly mode afterwards.